### DISCUSSION TITLE
RFC: Allow different Css Frameworks like Bootstrap5 RFC2021-04

### DISCUSSION THREAD

**Discussion by iJungleboy** [oqtane/oqtane.framework#1332] [Ideas]
## Background

Bootstrap5 has just been released. https://getbootstrap.com/

We must also be aware of the fact that CSS frameworks continuously improve to match browser capabilities (like flex-layouts), so this will not be the last Bootstrap and some day Bootstrap may not be the CSS framework of choice any more. 

As of now Oqtane is 100% tied to Bootstrap4 and this means a few things are very bootstrap 4 in a fairly rudimentary setup:

1. The initial skin preloads Bootstrap4 from a CDN - which is a trivial way to get started but isn't the way professional bootstrap themes are made. Good Bootstrap-Themes have some variables files which determine initial values, widths, gutters, breakpoints etc. and then run SASS to generate the final optimized bootstrap CSS.
2. Since the Oqtane UI-parts are running in the same page as anything custom (it's one single SPA after all) the bootstrap classes used in the default UI are not clearly separated from the styling in general. For example, creating a custom theme requires that this custom theme also has some styling which is copy-pasted from Oqtane, leading to redundancies and as Oqtane matures, themes must always be updated to match these styles.

## Goals

In this RFC I would like to achieve the following goals

1. Ensure that Oqtane can work with any CSS framework. As of now I would want to target Bootstrap 4 and 5 but the concept should be flexible enough that other frameworks will be possible (even if a bit harder to implement)

## Necessary Measures

1. Split CSS into concerns
2. Get all system dialogs / parts to be robust, so that they work even if the layout has major changes
3. Allow themes to load the built-in parts as needed, or supply their own as needed

## Implementation

For this I suggest the following

1. Separation of Concerns: The CSS should be split into logical **Parts**
    1. Layout CSS - responsible for the page layout, header, footer etc.
    2. In-Page Editing CSS responsible for everything which appears when you enter edit mode, incl. pop-out menus etc.  
       This is the most error-prone CSS as the layout could easily break this, so this would also be the part that would need the most love to ensure it's robust. 
    3. Dialogs which appear from the side or cover the entire screen
    4. Forms - which will probably have a lot of customization for each application
2. All built-in bits like module-edit-menus and Oqtane dialogs should have classes so that their styling can be different from the styles on the page. This is to ensure that changing the look of input fields doesn't break add-module dialogs etc. Or if it does break it, anybody can add more targeted CSS to the Oqtane dialogs so that they continue to work properly. 
3. Each theme should specify what **parts** it provides. This is to ensure that as the parts mature and new parts are added, Oqtane can supply the parts which the Theme doesn't even know about. 
4. _Optional_: Modules should be able to specify which parts it needs - so that only necessary parts are pre-loaded

## Steps to get there

I recommend the following:

1. Create separate solution in Oqtane for this, with SASS, Bootstrap etc. which has all the variables and stuff to build the correct parts.
2. Include these parts in the default distribution
3. With time, separate the concerns and create the standards/APIs to supply / request these parts
4. Experiment with a Bootstrap5 implementation to get it all to work in either worlds

Note: we've done a lot of similar stuff in DNN using Connect.Koi. https://connect-koi.net/ - this may be something to look at as well, but the Oqtane implementation could be very different. 

What do you all think?

**Comment by lanthonyneville** [comment]
I am a junior when it comes to CSS, so if this has already been considered, or does not make a contribution to the discussion, or is just plain ignorant or unfeasible, please feel free to say so! .....

It seems to me that part of the concern behind this conversation is the fact that, if the admin UI (i.e. everything that comes out of the box with Oqtane, admin menus, dialogs etc) is based on Bootstrap, then any CSS loaded by themes & modules that attaches to Bootstrap classes is in danger of distorting the admin UI. @iJungleboy is that what you are saying in your point 2? 

If so, could this whole issue be avoided by the admin UI having no dependency on Bootstrap (or any other popular framework that might be used by theme/module developers), and instead use a minimal set of CSS classes that are purely created for the admin UI, i.e. class names that have little danger of being referenced.

In my ignorance, it seems that the set of classes required for the admin UI should be substantially less than you find in a comprehensive framework like Bootstrap, e.g. maybe no need for the grid system?

Finally, if this is feasible, rather than hand creating the CSS for the admin UI, could it be taken from a more obscure CSS framework, or by "stealing" CSS from Bootstrap and renaming the classes, e.g. oq-modal-dialog instead of modal-dialog.

**Reply by iJungleboy** [reply]
@lanthonyneville This is partially correct - at least this is part of the problem. 

My understanding is that Bootstrap is hard-wired for 2 reasons

1. That the built-in admin-UIs already look pretty good without much effort, so that the Oqtane developers could focus on other priorities.
2. Because offering some kind of default / standard would make it easier for modules to be created - because when you install them they will look fairly good right from the start. 

Making the admins work independently of the page CSS would be an important goal - but that is certainly challenging and probably not on the next priorities of the Oqtane team. How it would be achieved is a question for the future. This could be using some minimal bootstrap which only affects Oqtane dialogs etc.

**Comment by sbwalker** [comment]
I am not opposed to adding this to the longer term roadmap.... perhaps after 3.0 (.NET6) is released might be a good time frame for consideration.

**Comment by ricardok1** [comment]
Hi @sbwalker 
Thanks for the excellent work.
I'm also on doubts to start with Oqtane due to CSS constrains.
I agree that you should show the example with your own choice but as your option not as "almost" mandatory.
Can you please update on the status / roadmap on this?
Thanks again!
Ricardo

**Reply by sbwalker** [reply]
@ricardok1 Just to clarify, Oqtane was upgraded to Bootstrap 5 last year - https://www.oqtane.org/blog/!/30/oqtane-2-2-upgrades-to-bootstrap-5 - and the admin modules were overhauled to use responsive form layouts. That being said, the admin modules and controls are still based on Bootstrap as a foundational CSS framework. Support for other CSS frameworks as outlined in this RFC is still up for future consideration on the roadmap... however there are other items such as support for MAUI which may be considered a higher priority by many.
